### PR TITLE
Prevent fields hiding joined fields

### LINKED ISSUE #29904: [MLv2] [Bug] Fields in `:joins` are supposed to be returned even if `:fields` is specified in parent level
I made a mistake in doing metadata calculation for queries that contain `:fields` and `:joins`: if `:fields`/`:breakout`/`:aggregation` is specified, only fields specified in those clauses are "returned", but this logic is actually incorrect:

```clj
(mapv
 :display_name
 (mt/cols (mt/run-mbql-query venues
            {:fields [$id
                      $name]
             :joins  [{:source-table $$categories
                       :alias        "Cat"
                       :condition    [:= $category_id &Cat.categories.id]
                       :fields       :all}]})))
;; =>
["ID" "Name" "Cat → ID" "Cat → Name"]
```

Currently MLv2 just returns

```clj
["ID" "NAME"]
```

We need to fix the `lib.metadata.calculation` logic in `metabase.lib.stage` so it matches the behavior of the QP.

### PR DESCRIPTION
**Pull request by metamben**
Fixes #29904.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30644)

### CODE REVIEW

File: src/metabase/lib/stage.cljc [hunk 1]
```diff
@@ -221,11 +221,23 @@
...
       ;; ... then calculate metadata for this stage
-      (or
-       (breakout-ags-fields-columns query stage-number unique-name-fn)
-       (lib.metadata.calculation/default-columns query stage-number (lib.util/query-stage query stage-number) unique-name-fn))))))
+      (cond
+        summary-cols
+        (into summary-cols field-cols)
+
+        field-cols
+        (do (doall field-cols)          ; force generation of unique names before join columns
```
Comment (metamben) [inline src/metabase/lib/stage.cljc:233]: This is quite ugly, but probably inherent in the nature of `unique-name-fn`.
(I don't like this side-effectful name generation. I would much more prefer a solution where we a global view on what columns we are generating the names for.)

File: test/metabase/lib/join_test.cljc [hunk 2]
```diff
@@ -214,7 +214,12 @@
               :lib/source-column-alias       "ID"
               :lib/desired-column-alias      "Cat__ID"
               :metabase.lib.field/join-alias "Cat"
-              :lib/source                    :source/fields}]
+              :lib/source                    :source/fields}
```
Comment (metamben) [inline test/metabase/lib/join_test.cljc:217]: This join field was added to the `:fields` clause.

File: test/metabase/lib/join_test.cljc [hunk 3]
```diff
@@ -214,7 +214,12 @@
...
+             {:name                          "NAME"
+              :lib/source-column-alias       "NAME"
+              :lib/desired-column-alias      "Cat__NAME"
+              :metabase.lib.field/join-alias "Cat"
+              :lib/source                    :source/joins}]
```
Comment (metamben) [inline test/metabase/lib/join_test.cljc:222]: This join field was not added to the `:fields` clause.

File: src/metabase/lib/stage.cljc (hunk 1 above)
Comment (snoe) [reply src/metabase/lib/stage.cljc:233]: Would it be possible to fill in `desired-column-alias` after metadata is gathered as a final step in `stage-metadata` etc? It doesn't seem like `unique-name-generator` is created in too many places.

**Review by snoe — APPROVED** [review]
It feels `fields` is stuck between trying to do too much or too little. Having a single source of truth for the next stage's (or final product's) `visible-columns` would be nice.

File: src/metabase/lib/stage.cljc (hunk 1 above)
Comment (metamben) [reply src/metabase/lib/stage.cljc:233]: I'll open an issue for this and see what @camsaul thinks.

File: src/metabase/lib/stage.cljc [hunk 4]
```diff
@@ -221,11 +221,23 @@
...
+            (into []
+                  (m/distinct-by #(dissoc % :source_alias :lib/source :lib/desired-column-alias))
```
[the remark this answers is not in the document]
Comment (metamben) [reply src/metabase/lib/stage.cljc:235]: There are existing tests that do this, so I assume the answer is yes. I found this confusing. @camsaul, WDYT?

**Pull request merged**